Fixed NodeThreadLeakTest

Donnerbart opened the pull request:
Hazelcast threads may need a little bit to shutdown, so we should use `assertJoinable()` to test them. The `ThreadLeakTestUtils` provide some convenient methods for that.

I checked that the test still fails if the `shutdown(true)` call in `NodeEngineImpl` is removed:
```java
        try {
            (...)
        } catch (Throwable e) {
            try {
                shutdown(true);
            } catch (Throwable ignored) {
                ignore(ignored);
            }
            throw rethrow(e);
        }
```

Fixes https://github.com/hazelcast/hazelcast/issues/12694

kwart commented on hazelcast/src/test/java/com/hazelcast/instance/NodeEngineThreadLeakTest.java:
```diff
@@ -38,18 +24,31 @@
 
 /**
- * Tests that exception in Node constructor leads to properly finished services.
+ * Tests that an exception in the {@link NodeEngineImpl} constructor leads to properly finished services.
```
The problem is/was in both constructors, could you use sth like `"the {@link NodeEngineImpl} and  {@link Node} constructors"`?

Donnerbart replied:
Oh, I didn't see the issues in `Node`. Fixed.

kwart commented on hazelcast/src/test/java/com/hazelcast/instance/NodeEngineThreadLeakTest.java:
```diff
@@ -38,18 +24,31 @@
+@PrepareForTest({DefaultNodeContext.class, NodeEngineImpl.class})
 @Category(QuickTest.class)
-public class NodeThreadLeakTest {
+public class NodeEngineThreadLeakTest extends HazelcastTestSupport {
```
Renaming is not necessary here. The ur-issue was not in the `NodeEngine`, but in call `nodeContext.createConnectionManager(this, serverSocketChannel)` within the `Node.<init>`. It fails when the SSL/TLS is not configured properly.

Donnerbart commented:
I've reverted the naming and added `Node` + `NodeEngineImpl` to the JavaDoc.

The `@BeforeClass` should not be needed, since the methods in `ThreadLeakTestUtils` filter out the threads, which are captured at the beginning:
```java
    private static Thread[] getJoinableThreads(Set<Thread> oldThreads, Set<Thread> newThreads) {
        Set<Thread> diff = new HashSet<Thread>(newThreads);
        diff.removeAll(oldThreads);
        diff.remove(Thread.currentThread());
        removeWhitelistedThreadClasses(diff);

        Thread[] joinable = new Thread[diff.size()];
        diff.toArray(joinable);
        return joinable;
    }
```
The current thread and `JitterThread` are also filtered out. This seems to work fine in the `ThreadLeakTest` and `ThreadLeakClientTest` (which check the same thing without startup failures).
